Add github action to push docker images with latest tag.

revit13 opened the pull request:
## Why are these changes needed?
This PR automates the process of pushing Docker images with `latest` tag upon each merge to the development branch. It accomplishes this by storing the images generated during the execution of the `test-image` target using the `https://github.com/actions/upload-artifact` action and subsequently downloading and pushing them to the repository following the completion of testing.
It also changes ray docker images to have a base image and thus avoids the need to call update_dockerfile.sh script to update the ray version in set-versions.

the pushed images rae:
```bash
quay.io/dataprep1/data-prep-kit/proglang_select-ray:latest
quay.io/dataprep1/data-prep-kit/proglang_select-python:latest
quay.io/dataprep1/data-prep-kit/malware-ray:latest
quay.io/dataprep1/data-prep-kit/malware-python:latest
quay.io/dataprep1/data-prep-kit/code_quality-ray:latest
quay.io/dataprep1/data-prep-kit/code_quality-python:latest
quay.io/dataprep1/data-prep-kit/code2parquet-ray:latest
quay.io/dataprep1/data-prep-kit/code2parquet-python:latest
quay.io/dataprep1/data-prep-kit/lang_id-ray::latest
quay.io/dataprep1/data-prep-kit/tokenization-ray:latest
quay.io/dataprep1/data-prep-kit/tokenization-python:latest
quay.io/dataprep1/data-prep-kit/resize-ray:latest
quay.io/dataprep1/data-prep-kit/resize-python:latest
quay.io/dataprep1/data-prep-kit/profiler-ray:latest
quay.io/dataprep1/data-prep-kit/noop-spark:latest
quay.io/dataprep1/data-prep-kit/noop-ray:latest
quay.io/dataprep1/data-prep-kit/noop-python:latest
quay.io/dataprep1/data-prep-kit/filter-spark:latest
quay.io/dataprep1/data-prep-kit/filter-ray:latest
quay.io/dataprep1/data-prep-kit/filter-python:latest
quay.io/dataprep1/data-prep-kit/fdedup-ray:latest
quay.io/dataprep1/data-prep-kit/ededup-ray:latest
quay.io/dataprep1/data-prep-kit/doc_id-spark:latest
quay.io/dataprep1/data-prep-kit/doc_id-ray:latest
quay.io/dataprep1/data-prep-kit/data-prep-kit-spark-3.5.1:latest
quay.io/dataprep1/data-prep-kit/lang_id-python:latest
quay.io/dataprep1/data-prep-kit/lang_id-ray:latest
quay.io/dataprep1/data-prep-kit/kfp-data-processing:latest
```

## Related issue number (if any).

https://github.com/IBM/data-prep-kit/issues/176

revit13 committed:
Add github action to push docker images with latest tag.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Merge branch 'dev' into github13

revit13 committed:
Debug workflow.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
fix.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Handle latest tag.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Fixes after testing.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Merge branch 'dev' into github13

revit13 committed:
Fixes after testing.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Add base image to ray images.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Fix final stage.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Minor change.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Test publish phase.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Trgger push only     on merge to dev.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Test push images.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Minor change.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Fixes after testsing.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Minor fix.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Address review comments.

Signed-off-by: Revital Sur <eres@il.ibm.com>

roytman commented on scripts/transforms/update_workflow_tags.sh:
```diff
@@ -1,15 +1,17 @@
+if [[ $# != 3 ]]; then
   cat << EOF
-	"Incorrect number of parameters provided. The required parameters are versions_file and pipeline_path. 
+	$1 $2 
```
should we add more explanation here? instead just $1 and $2

revit13 replied:
Thanks I removed the redundant printing...

roytman commented on tools/ingest2parquet/Dockerfile:
```diff
@@ -1,4 +1,6 @@
-FROM docker.io/rayproject/ray:2.24.0-py310
+ARG BASE_IMAGE=docker.io/rayproject/ray:2.24.0-py310
+
+FROM ${BASE_IMAGE}
```
wow, that's very suitable.

roytman commented on .make.defaults:
```diff
@@ -179,6 +181,16 @@ __check_defined = \
+	-rm -rf $(REPOROOT)/artifacts > /dev/null 2>&1
+
+.PHONY: .defaults.docker-save-image
+.defaults.docker-save-image:
```
should we add a rule to the `make clean` to clean up the ARTIFACTS_DIR, the images can take up a lot of space.

revit13 replied:
Yes the removal of ARTIFACTS_DIR is in make clean target.

roytman reviewed (approved):
LGTM

daw3rd reviewed (changes requested):
@revit13 and I spoke and agreed we will try and simplify things by always publishing the "latest" docker images if a) it is a merge into dev and b) the make test-image passes.
Publishing will be done after `make test-image` only for dev merges and the tag in this case will be `latest`.  The "releases" will be handled manually (for now?) and publishing will be done out of `releases/vX.Y.Z` branch.  But perhaps this can somehow be automated as well?  Maybe not using `latest` tag?

revit13 committed:
Address review comments.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Minor fix.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 committed:
Minor fix.

Signed-off-by: Revital Sur <eres@il.ibm.com>

revit13 commented:
> @revit13 and I spoke and agreed we will try and simplify things by always publishing the "latest" docker images if a) it is a merge into dev and b) the make test-image passes. Publishing will be done after `make test-image` only for dev merges and the tag in this case will be `latest`. The "releases" will be handled manually (for now?) and publishing will be done out of `releases/vX.Y.Z` branch. But perhaps this can somehow be automated as well? Maybe not using `latest` tag?

@daw3rd I changed the action to push as described above. Please let me know if thats OK. I also added a condition to push the images upon creating a new tag which should cover publishing for release. Thanks

revit13 committed:
Add a check for DOCKER_IMAGE_NAME

Signed-off-by: Revital Sur <eres@il.ibm.com>

daw3rd commented on transforms/code/code2parquet/ray/Dockerfile:
```diff
@@ -1,4 +1,6 @@
-FROM docker.io/rayproject/ray:2.24.0-py310
+ARG BASE_IMAGE=docker.io/rayproject/ray:2.24.0-py310
```
How might this effect changing of RAY_VERSION in .make.versions?  That is, if that is changed, what user of this docker file needs to changes what its passing in here.  I suppose this is really a question, what is this used for?

revit13 replied:
In .make.defaults we define: `RAY_BASE_IMAGE?=docker.io/rayproject/ray:${RAY}-py310` and in each ray version MAKEFILE of the transforms `BASE_IMAGE=${RAY_BASE_IMAGE}` is defined such that the base image is updated based on RAY version.

daw3rd commented on .make.defaults:
```diff
@@ -179,6 +181,18 @@ __check_defined = \
+.PHONY: .defaults.docker-save-image
+.defaults.docker-save-image:
+	@# Help: Save docker image as tar file to ${ARTIFACTS_DIR} directory.
+	mkdir -p ${ARTIFACTS_DIR}/
```
Now that we're trying to add checks, can you add a check for DOCKER_IMAGE_NAME.  Maybe this (and others) goes at the top of this file though instead of at each referencing target?  I guess its the "global" ones like this that could go at the top of this file?

revit13 replied:
Checking `DOCKER_IMAGE_NAME` in the top of this file results in an error as its is set in the Makefiles that generate Docker images and include .make.defauls, for example  in `.make.transforms`. So I added it to `image` and `publish` targets.

daw3rd commented on .make.versions:
```diff
@@ -23,14 +23,21 @@ DPK_VERSION_SUFFIX=.dev0
 
-KFP_DOCKER_VERSION=$(DPK_VERSION)
-KFP_DOCKER_VERSION_v2=$(DPK_VERSION)
+KFP_DOCKER_VERSION=$(DOCKER_IMAGE_VERSION)
```
Maybe not now, but should be KFP_DOCKER_VERSION_V1?

daw3rd reviewed (approved):
lgtm
